### PR TITLE
Feat: V1

### PR DESCRIPTION
**Pull request by Keyrxng**
Resolves https://github.com/ubiquibot/plugins-wishlist/issues/35

### CODE REVIEW

**Comment by Keyrxng** [comment]
> > I'm thinking that we should also build it so it's reusable for other partners too and not just for our specific use-case

> Yes

@rndquu any suggestions on slash commands and/or how to make it more reusable?

(admin only)
- `/faucet keyrxng <chain_id> <amount> <token>`: ERC20 and native
- `/faucet keyrxng <chain_id> <amount>` native only
- `/faucet keyrxng <chain_id>` native only

(anyone)
- `/register` creates a storage entry and comments them to go to our registration portal

In the portal we will create a new entry if one doesn't exist so they can fully register via the UI with github login.

So I went ahead and used an array of networkIds and just try to transfer on each network. This could probably have better handling but in our case we control the underlying chain that the funds exist on so we'd probably only use gnosis right?

**Comment by rndquu** [comment]
> I have not manually QA'd 

:(

> saving a json in ubiquibot-config it should be inspected to see

There's too much for a single PR. JSON storage should be solved as a part of https://github.com/ubiquibot/plugin-template/issues/2 and added to the SDK at https://github.com/ubiquibot/sdk. Keeping JSON storage related code in the faucet plugin is not "reusable".

> `/faucet keyrxng <chain_id> <amount> <token>`

Why do we need the `/faucet` command? I actually though the plugin will fund addresses automatically without any commands.

**Comment by Keyrxng** [comment]
> should be solved as a part of https://github.com/ubiquibot/plugin-template/issues/2 and added to the SDK at https://github.com/ubiquibot/sdk. Keeping JSON storage related code in the faucet plugin is not "reusable".

My intention was to implement it here as no other attempts at it exist (as far as I'm aware) and then have it standardized through this plugin and build into the SDK or `plugin-template` wherever is best. I created a branch in my `start-stop` fork a while back to implement it but didn't see it through.

The alternative is for me to use our Supabase so we have our users and then update the DB or use on-chain to tracks transfers from the funding address to the user address or something similar

Let me know which I should follow through for V1 and I will

> Why do we need the /faucet command? I actually thought the plugin will fund addresses automatically without any commands.

I guess we do not need any other commands. I figured it would be beneficial to be able to invoke the faucet directly and have it be reusable as more than just a gas subsidy on `issues.closed`. I could remove it no problem but would it also make sense to remove the ERC20 distribution too since we already have rewards issued on `issues.closed`?

If restricting things to be just a gas subsidy plugin we should name it something along those lines instead of `crypto-faucet` as I think that's misleading.

---

Maybe what I'm thinking could have it's own dedicated plugin? Slash command(s) which allow admins (or anyone registered through AA) to easily transfer ERC20/native tokens from their own wallet or org config specified wallet(s). It has many use-cases but a few are:

- manual payments made super easy
- they could be made to render as permits for easy auditing
- off-the-cuff bonuses
- consolation rewards for time invested in tasks which get closed for reasons unrelated to the contributor's work

**Comment by whilefoo** [comment]
> My intention was to implement it here as no other attempts at it exist (as far as I'm aware) and then have it standardized through this plugin and build into the SDK or `plugin-template` wherever is best. I created a branch in my `start-stop` fork a while back to implement it but didn't see it through.

There's no consensus as to what to use. The wallets are currently still stored in the Supabase so I think we should use that for v1.

@Keyrxng why does the user need to go to `safe.ubq.fi` to generate via webauthn instead of just putting their own wallet address. Is it to prevent abuse or is there something else I'm not aware of?

**Comment by 0x4007** [comment]
I really don't like using Supabase so any effort to implement JSON storage on git I fully support. If you think its easier to implement here first, please do. Afterwards we can use the logic inside of the SDK and slowly remove the code from within the plugin if it makes sense to. It's okay if we have redundant logic as well, given that plugins are meant to be sort of "one offs"

**Comment by Keyrxng** [comment]
> why does the user need to go to safe.ubq.fi to generate via webauthn instead of just putting their own wallet address. Is it to prevent abuse or is there something else I'm not aware of?

@whilefoo it's because we are creating the EOA that'll we'll bind to their account abstraction profile and we do this through a mixture of OAuth and passkeys which are browser-based

---

Okay so if V1 runs on Supabase and the `/register` command just tells them to go to the portal to register:

1. We need to update the db to accept another column like "subsidized" so we can track subsidies unless using onchain
2. As we don't know the EOA at this point, should this plugin create a user and wallet entry with a `null` wallet address and it will be updated in the UI?
3. On `issue.closed` this plugin will fetch the user entry and if `!subsidized` then we gift gas.

As we plan only to subsidize once and only with gas I'll remove the ERC20 and `howManyTimesUserCanClaim` configurables

**Comment by whilefoo** [comment]
> @whilefoo it's because we are creating the EOA that'll we'll bind to their account abstraction profile and we do this through a mixture of OAuth and passkeys which are browser-based

Is that meant to be used by developers who don't necessarily know how to use crypto or not have Metamask? Or is there another reason to this approach?

> 1. We need to update the db to accept another column like "subsidized" so we can track subsidies unless using onchain

that's correct

> 2\. As we don't know the EOA at this point, should this plugin create a user and wallet entry with a `null` wallet address and it will be updated in the UI?

this is a bit messy because if we create a wallet record that's set to null, it might break other plugins which don't expect null. Also if the user doesn't complete the registration process then the record will remain null forever.

> 3\. On `issue.closed` this plugin will fetch the user entry and if `!subsidized` then we gift gas.

basically I think we will need two column, one on the user `subsidized` which means this user has been or has not been subsizided and one for the wallet if wallet was created with `/register` command, but that will be tricky to track.

if we modify `safe.ubq.fi` so that the user needs to login with Github and then generate a EOA with a passkey, we can create a wallet record at that point, or am I missing something?

why do we only offer subsidy if the wallet is created with `safe.ubq.fi`?

**Comment by Keyrxng** [comment]
> Is that meant to be used by developers who don't necessarily know how to use crypto or not have Metamask? Or is there another reason to this approach?

That's one reason some others are streamlining the user experience, making it possible to claim with just your personal device not necessarily a device with your wallet tied to it.

> this is a bit messy because if we create a wallet record that's set to null, it might break other plugins which don't expect null. Also if the user doesn't complete the registration process then the record will remain null forever.

I was thinking this too. So really in V1 `/register` will just tell them to go to the portal in order to register with maybe some other additional info. "This is a beta feature... This is what to do..."

> if we modify safe.ubq.fi so that the user needs to login with Github and then generate a EOA with a passkey, we can create a wallet record at that point, or am I missing something?

If in V1 we are using only our Supabase then updating the DB on `safe.ubq.fi` is no problem but V2 it will be although we can deal with that later.

> why do we only offer subsidy if the wallet is created with safe.ubq.fi?

We plan to offer it to any new contributor I believe but the focus was on the AA side more-so than those with an EOA already but both are subsidized once.

---

1. Update DB to capture what's needed
2. Update plugin so that `/register` only redirects them to the portal
3. Generate EOA and store user details
4. Prompt user to return to the issue and use the `/start` command to begin the task
5. On `issue.closed` we simply look at `subsidy` field and distribute once regardless of registration method (and regardless of balance? Everyone gets one at least once, or should we balance check them?)

**Comment by whilefoo** [comment]
> 5\. On `issue.closed` we simply look at `subsidy` field and distribute once regardless of registration method (and regardless of balance? Everyone gets one at least once, or should we balance check them?)

Ideally we should the balance is 0

**Comment by 0x4007** [comment]
Yeah maybe we can just infinite subsidy. We don't need to keep track of a user's state whether they received a faucet subsidy in the past. Because it's incredibly difficult to make one's account balance perfectly zero after you added any gas.

**Comment by Keyrxng** [comment]
> Yeah maybe we can just infinite subsidy. We don't need to keep track of a user's state whether they received a faucet subsidy in the past. Because it's incredibly difficult to make one's account balance perfectly zero after you added any gas.
> 
> So this seems very simple to always check everybody's gas balance of their registered wallet. And if it's perfectly 0 then we can send enough for a few transactions.
> 
> It would be useful to also teach the user how to unwrap their WXDAI from their reward. That way they should never have a gas problem. Perhaps in the future, we can append a message to their permit link when a faucet subsidy occurs, which can render as a banner on the rewards UI for their first claim.

> It would be useful to also teach the user how to unwrap their WXDAI from their reward.

image

Perhaps we create another discussion for that topic if it makes sense to do it that way as the first one seems to be effective

So no need to update the DB then we just rely on a exact 0 balance and we'll distribute so long as they have registered their wallet to the DB. This technically could happen multiple times but it's okay as it's less than pennies at a time that we distro and it's sort of difficult to obtain a 0 balance

**Comment by 0x4007** [comment]
Well if we can also check the GitHub user id then it can make it not worth to farm our faucet (they would have to create a new GitHub every time) 

This is intended for first time contributors, not for people who already know what's going on.

File: .github/workflows/jest-testing.yml [hunk 1]
```diff
@@ -19,8 +19,29 @@ jobs:
...
+          for i in {1..30}
+          do
+            if curl -s http://localhost:8545; then
+              break
+            fi
+            sleep 1
+          done || exit 1
+
       - name: Jest With Coverage
-        run: yarn install --immutable --immutable-cache --check-cache && yarn test
```
Comment (gentlementlegen) [inline .github/workflows/jest-testing.yml:23]: Why removing these?

File: package.json [hunk 2]
```diff
@@ -32,8 +33,10 @@
     "@octokit/rest": "20.1.1",
     "@octokit/webhooks": "13.2.7",
     "@sinclair/typebox": "0.32.33",
+    "@ubiquity-dao/rpc-handler": "^1.2.3",
     "@ubiquity-dao/ubiquibot-logger": "^1.3.0",
     "dotenv": "16.4.5",
+    "ethers": "5.7.2",
```
Comment (gentlementlegen) [inline package.json:39]: So we won't use `v6` for this one? Any difference?

Comment (0x4007) [reply package.json:39]: ChatGPT knows v5

File: .github/workflows/jest-testing.yml (hunk 1 above)
Comment (Keyrxng) [reply .github/workflows/jest-testing.yml:23]: `yarn install` happen a few steps above so this only needs to run `yarn test`. I no longer use the flags as I assumed they were only relevant to the `Jest With Coverage` step, is this correct?

File: package.json (hunk 2 above)
Comment (Keyrxng) [reply package.json:39]: I think the solution to approaching a migration to `v6` would be to upgrade the `rpc-handler` and install both `v5` and `v6` then as an init value tell it which version to return. 

We default to `v5` as to avoid breaking anything and in newer projects like this it can be put to the test?

Comment (Keyrxng) [reply package.json:39]: I'll resolve this for now as I've opened an issue for it and we can circle back to it
